### PR TITLE
NetBird SSH

### PR DESCRIPTION
**Pull request by braginini**

### CODE REVIEW

**Comment by braginini** [comment]
If an agent has been upgraded from any version without SSH support, all others won't receive an update indicating that SSH is now possible to this peer.
Should be fixed

File: client/internal/config.go [hunk 1]
```diff
@@ -145,7 +158,7 @@ func GetConfig(managementURL, adminURL, configPath, preSharedKey string) (*Confi
 	} else {
 		// don't overwrite pre-shared key if we receive asterisks from UI
 		pk := &preSharedKey
-		if preSharedKey == "**********" {
+		if preSharedKey == "**********" || preSharedKey == "" {
```
Comment (mlsmaycon) [inline client/internal/config.go]: we shouldn't dismiss empty string as sometimes user wants to remove a preshared key from its config

File: client/internal/engine.go [hunk 2]
```diff
@@ -414,6 +420,41 @@ func (e *Engine) handleSync(update *mgmProto.SyncResponse) error {
...
 }
 
+func (e *Engine) updateSSH(sshConf *mgmProto.SSHConfig) error {
+	if sshConf.GetSshEnabled() {
+		// start SSH server if it wasn't running
+		if e.sshServer == nil {
+			//nil sshServer means it has not yet been started
+			var err error
+			e.sshServer, err = e.sshServerFunc(e.config.SSHKey,
+				fmt.Sprintf("%s:%d", e.wgInterface.Address.IP.String(), 2222))
```
Comment (mlsmaycon) [inline client/internal/engine.go]: port should come from config for cases when port 2222 is already in use

File: client/internal/engine.go [hunk 3]
```diff
@@ -414,6 +420,41 @@ func (e *Engine) handleSync(update *mgmProto.SyncResponse) error {
...
+			if err != nil {
+				return err
+			}
+			go func() {
```
Comment (mlsmaycon) [inline client/internal/engine.go:440]: In the event of an error (e.g. port in use), we should clean the e.sshServer

File: client/internal/engine_test.go [hunk 4]
```diff
@@ -40,6 +43,125 @@ var (
...
+	key, err := wgtypes.GeneratePrivateKey()
+	if err != nil {
+		t.Fatal(err)
+		return
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	engine := NewEngine(ctx, cancel, &signal.MockClient{}, &mgmt.MockClient{}, &EngineConfig{
```
Comment (mlsmaycon) [inline client/internal/engine_test.go:61]: lets avoid using same interface configuration between tests

File: client/internal/engine.go [hunk 5]
```diff
@@ -54,6 +55,8 @@ type EngineConfig struct {
 
 	// UDPMuxSrflxPort default value 0 - the system will pick an available port
 	UDPMuxSrflxPort int
+
+	SSHKey []byte
```
Comment (mlsmaycon) [inline client/internal/engine.go:61]: We should distinguish private and public ssh keys, here we store private key but other methods use sshKey for public SSH Key (e.g. Login)

File: client/ssh/server.go [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+func getShellType() string {
+	shell := os.Getenv("SHELL")
+	if shell == "" {
+		shell = "sh"
+	}
+	return shell
+}
+
+// sessionHandler handles SSH session post auth
+func (srv *DefaultServer) sessionHandler(s ssh.Session) {
```
Comment (mlsmaycon) [inline client/ssh/server.go]: let's avoid using single-letter input variables

File: client/ssh/server.go [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+			return
+		}
+		err = s.Exit(1)
+		if err != nil {
+			return
+		}
+	}
+}
+
+func (srv *DefaultServer) stdInOut(f *os.File, s ssh.Session) {
```
Comment (mlsmaycon) [inline client/ssh/server.go]: let's avoid using single-letter input variables

File: client/ssh/util.go [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+const ECDSA KeyType = "ecdsa"
+
+// RSA is key of type rsa
+const RSA KeyType = "rsa"
+
+// RSAKeySize is a size of newly generated RSA key
+const RSAKeySize = 2048
+
+// GeneratePrivateKey creates RSA Private Key of specified byte size
+func GeneratePrivateKey(t KeyType) ([]byte, error) {
```
Comment (mlsmaycon) [inline client/ssh/util.go]: let's avoid using single-letter input variables

File: client/ssh/window_unix.go [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+
+package ssh
+
+import (
+	"os"
+	"syscall"
+	"unsafe"
+)
+
+func setWinSize(f *os.File, w, h int) {
```
Comment (mlsmaycon) [inline client/ssh/window_unix.go]: let's avoid using single-letter input variables

File: client/ssh/window_windows.go [hunk 10]
```diff
@@ -0,0 +1,9 @@
+package ssh
+
+import (
+	"os"
+)
+
+func setWinSize(f *os.File, w, h int) {
```
Comment (mlsmaycon) [inline client/ssh/window_windows.go]: let's avoid using single-letter input variables

File: management/server/grpcserver.go [hunk 11]
```diff
@@ -290,6 +296,19 @@ func (s *Server) Login(ctx context.Context, req *proto.EncryptedMessage) (*proto
 			return nil, status.Error(codes.Internal, "internal server error")
 		}
 	}
+
```
Comment (mlsmaycon) [inline management/server/grpcserver.go:299]: can't put his comment in line 273, but should we use the peer key from the PeerKeys payload?

File: management/server/grpcserver.go [hunk 12]
```diff
@@ -365,7 +384,8 @@ func toWiretrusteeConfig(config *Config, turnCredentials *TURNCredentials) *prot
 
 func toPeerConfig(peer *Peer) *proto.PeerConfig {
 	return &proto.PeerConfig{
-		Address: fmt.Sprintf("%s/%d", peer.IP.String(), SubnetSize), // take it from the network
+		Address:   fmt.Sprintf("%s/%d", peer.IP.String(), SubnetSize), // take it from the network
+		SshConfig: &proto.SSHConfig{SshEnabled: true},                 //TODO REMOVE THIS HARDCODED VALUE
```
Comment (mlsmaycon) [inline management/server/grpcserver.go]: remove hardcoded flag

**Review by mlsmaycon — CHANGES_REQUESTED** [review]

**Review by mlsmaycon — APPROVED** [review]

**Pull request merged**